---
title: "runDRIM+DEXseq"
output: 
  html_document:
    df_print: paged
---

N.b. this script must be run in the "R-rstudio" environment:
"/shared/sudlab1/General/apps/conda/conda-install/envs/R-rstudio"

Import necessary libraries

```{r import_libs, warning=FALSE }
library(tidyverse, quietly=T)
library(tximport, quietly=T)
library(DESeq2, quietly=T)
library(stringr, quietly=T)
library(ggplot2)
#library(GenomicFeatures)
library(DRIMSeq)
library(stageR)
#library(DEXSeq)
#library(tidyr)
library(data.table)
```

Load in the configs / run-specific information

```{r load_configs}


#get formula
formula = "~var1"

#we need to know what var1 values to filter by
filter_by_var1 = c("siDSRed", "siUPF1")



```

Generate sample table for tximport

```{r sample_table_from_sf}
#list all salmon files in quantification directory

quant_dir = "../../../../../HCT-NMD_KD/requant/quantification.dir/"

all_files = list.files(quant_dir, pattern="*")
all_files = file.path(quant_dir,  all_files, "quant.sf")

sample_names <- gsub(".star.agg-agg-agg.filtered", "", basename(dirname(all_files)))
names(all_files) <- sample_names

sample_table = data.frame(files=all_files, sample_id=sample_names)
sample_table = sample_table %>% mutate(var1=ifelse(stringr::str_detect(sample_id, "HCT116-R"), "siDSRed", "siUPF1"))
sample_table$files = as.character(sample_table$files)
```

How many samples are in each condition?

```{r overview_table}
table(sample_table$var1)
```

Check the files exist

```{r check_files_exist}
all(file.exists(sample_table$files))
```
 
Created named list of files

```{r named_list_of_files}
files = sample_table$files
names(files) = sample_table$sample_id
head(files)
```

Import with tximport

```{r tximport}
txi = tximport(files, type="salmon", txOut=TRUE, countsFromAbundance="no")
cts = txi$counts
#remove any rows where there are no counts
cts = cts[rowSums(cts) > 0,]
```

### Transcript to gene mapping

Load tx2gene from csvdb_files and rename so it matches quantification convention

```{r load_txdf}
txdf = read.delim("../../../../final_classifications/filtered_genesets.dir/TCGA-agg-agg.filtered.tx2gene.tsv")
txdf = txdf %>% transmute(GENEID = match_gene_id, TXNAME = transcript_id)
tab = table(txdf$GENEID)
txdf$ntx = tab[match(txdf$GENEID, names(tab))]
```

We need nrows to be same for both the txdf and cts so we need to manipulate both (just in case they differ, which they likely will now due to having filtered out transcripts with count == 0)

```{r reduce_txdf}
nrow(txdf)
nrow(cts)

cts.names=row.names(cts)
in_cts_not_txdf = base::setdiff(cts.names, txdf$TXNAME)
in_txdf_not_cts = base::setdiff(txdf$TXNAME, cts.names)

cts = cts[!(row.names(cts) %in% in_cts_not_txdf),]
txdf = txdf %>% filter(!TXNAME %in% in_txdf_not_cts)

nrow(txdf)
nrow(cts)
```

Now they should be the same!

### DRIMSeq

Create counts matrix with each row as a transcript, and each column as a sample

```{r count_matrix}
library(data.table)
counts = data.frame(cts)
counts = setDT(counts, keep.rownames = TRUE)
colnames(counts)[which(names(counts) == "rn")] = "feature_id"
counts = merge(counts, txdf, by.x = "feature_id", by.y = "TXNAME")
counts = counts[,1:(length(counts)-2)]
counts = data.frame(counts)
counts = counts %>% relocate(GENEID) 
colnames(counts)[which(names(counts) == "GENEID")] = "gene_id"
row.names(counts) = counts$feature_id
```

For some reason this has changed colnames from identifier-var1-replicate to identifier.var1.replicate. So lets change both to replace all dots with dashes.

```{r rename_count_matrix_heads}
names(counts) = stringr::str_replace_all(names(counts), "\\.", "\\-")
sample_table = sample_table %>% mutate(sample_id = stringr::str_replace_all(sample_id, "\\.", "\\-"))
```

Additionally, there appear to be a few MSTRG transcripts that have no associated gene... so filter those away

```{r filter_novel_no_genes}
counts = counts %>% filter(gene_id != "")
```

Create dmDSdata object

```{r create_dmDSdataObject}
d = dmDSdata(counts=counts, samples=sample_table)
d
```

```{r head_dmDSdataObject}
counts(d[1,])
```

DRIM filtering

```{r DTU_filtering}
n = dim(counts)[2]-2
tx_n = 0.4
tx_expr = 10
tx_prop = 0.1
gene_n = 0.7
gene_expr = 10

d = dmFilter(d,
             min_samps_feature_expr=tx_n, min_feature_expr=tx_expr,
             min_samps_feature_prop=tx_n, min_feature_prop=tx_prop,
             min_samps_gene_expr=(gene_n), min_gene_expr=gene_expr)
d
```

How many isoforms are there for each gene?

```{r table_d}
table(table(counts(d)$gene_id))
```

Now to create the design matrix based on design_tsv complexity 

```{r design_formula}
if(formula == "~var1"){
  design_full = model.matrix(~var1, data=DRIMSeq::samples(d))
}else if(formula == "~var1 + var2 + var1:var2"){
  design_full = model.matrix(~var1 + var2 + var1:var2, data=DRIMSeq::sample(d))
}
colnames(design_full)
```

Run DRIMSeq

```{r run_DRIM_precision}
d = dmPrecision(d, design=design_full, BPPARAM = BiocParallel::bpparam())
```
```{r run_DRIM_fit}
d = dmFit(d, design=design_full, BPPARAM = BiocParallel::bpparam())
```
```{r run_DRIM_test}
d = dmTest(d, coef=colnames(design_full)[2], BPPARAM = BiocParallel::bpparam())
```

Results

```{r result_DRIM}
res = DRIMSeq::results(d)
head(res)
```

Now at the transcript level

```{r tx_level_DRIM}
res.txp = DRIMSeq::results(d, level="feature")
head(res.txp)
```

We want to filter this by p-value
but we also need to make sure that any p-values which return an na value are set to 1 and not 0

```{r filter_drim_pval}
no.na = function(x) ifelse(is.na(x), 1, x)
res$pvalue = no.na(res$pvalue)
res.txp$pvalue = no.na(res.txp$pvalue)
```

We can plot the top (most confident, i.e. lowest p-value) switch

```{r plot_drim_pval}
idx = which(res$adj_pvalue < 0.05)[1]
res[idx,]
plotProportions(d, res$gene_id[idx], "var1")
```

### Final processing of DRIMSeq result with stageR

```{r stageR_drim}
pScreen = res$pvalue

names(pScreen) = res$gene_id

pConfirmation = matrix(res.txp$pvalue, ncol=1)
rownames(pConfirmation) = res.txp$feature_id

tx2gene = res.txp[,c("feature_id", "gene_id")]

stageRObj = stageRTx(pScreen = pScreen,
                     pConfirmation = pConfirmation,
                     pScreenAdjusted = FALSE,
                     tx2gene = tx2gene)

stageRObj = stageWiseAdjustment(stageRObj, method="dtu", alpha=0.05)

suppressWarnings({
  drim.padj = getAdjustedPValues(stageRObj, order=FALSE, onlySignificantGenes = FALSE)
})

head(drim.padj)
```


Now get the fract_expr data
```{r}
txdf = read.delim("../../../../final_classifications/filtered_genesets.dir/TCGA-agg-agg.filtered.tx2gene.tsv") %>%
  select(transcript_id=transcript_id, gene_id=match_gene_id)
# txi_gene <- tximport::summarizeToGene(txi, txdf) 
# 
# gene_tpm <- txi_gene$abundance
# colnames(gene_tpm) <- sample_table$sample_id
# 
# as.data.frame(gene_tpm) %>% 
#   rownames_to_column("gene_id") %>%
#   pivot_longer(-gene_id, names_to="sample_id", values_to="gene_tpm") %>%
#   filter(gene_tpm>0) -> gene_tpm

tx_tpm <- txi$abundance
colnames(tx_tpm) <- sample_table$sample_id

as.data.frame(tx_tpm) %>% 
  rownames_to_column("transcript_id") %>%
  pivot_longer(-transcript_id, names_to="sample_id", values_to="transcript_tpm") %>%
  left_join(sample_table) %>%
  group_by(transcript_id, var1) %>%
  summarize(transcript_tpm = median(transcript_tpm, na.rm=T)) %>%
    left_join(txdf) %>%
  group_by(gene_id, var1)  %>%
  mutate(gene_tpm=sum(transcript_tpm),
         fract_expr = (transcript_tpm+0.1)/(gene_tpm+0.1)) %>%
  pivot_wider(id_cols=transcript_id,
              names_from=var1,
              names_sep="_",
              values_from=c("fract_expr", "transcript_tpm"),
              ) -> fract_expr

all_utrons <- read.delim("../../../../final_classifications/utron_beds.dir/TCGA-agg-agg.all_utrons.ids.gz", header=F)
novel_utrons <- read.delim("../../../../final_classifications/utron_beds.dir/TCGA-agg-agg.novel_utrons.ids.gz", header=F)
partnered_utrons <- read.delim("../../../../final_classifications/utron_beds.dir/TCGA-agg-agg.partnered_utrons.ids.gz", header=F)
non_cds <- read.delim("../../../../final_classifications/utron_beds.dir/TCGA-agg-agg.no_cds_utrons.ids.gz", header=F)


dtu_results_table <-
  drim.padj %>% 
  select(gene_id=geneID, transcript_id =txID, gene_padj=gene, transcript_padj=transcript) %>% 
  left_join(fract_expr) %>%
  mutate(has_3UI = transcript_id %in% all_utrons$V1,
         has_novel_3UI = transcript_id %in% novel_utrons$V1,
         has_e3UI = transcript_id %in% 
           non_cds$V1) 

write.table(dtu_results_table, "Supplementary_data_10.tsv", sep="\t", row.names=F, quote=F)
```

```{r, fig.width=3, fig.height=2}
dtu_results_table %>%
  filter(transcript_tpm_siDSRed >2.5 | transcript_tpm_siUPF1 >2.5) %>%
  filter(has_3UI==T,
         !is.na(transcript_padj)) %>%
 ggplot() + aes(x=fract_expr_siDSRed, y=fract_expr_siUPF1, color=transcript_padj < 0.05, alpha=transcript_padj < 0.05) +
  geom_point(size=0.5) +
  theme_bw(base_size=8) +  xlab("Tx/Gene in Control") +
  ylab("Tx/Gene in siUPF1") +
  coord_fixed() +
  theme(axis.text = element_text(size=8),
        aspect.ratio=0.75,
        legend.position = "none") + 
  scale_color_manual(values=c("black", "red")) + 
  scale_alpha_manual(values=c(0.5,1))  -> saveme
png("siUPF1_DTUtrons.png", height=2, width=3, units="in", res=1200)
saveme
dev.off()
```

